cmd,interfaces: add initial support for Solus

ikeydoherty opened the pull request:
- Do not use re-exec on Solus
 - Allow autogen to pass `$@` arguments to `configure` script
 - Recognise Solus as a "classic" distribution
 - Support usr-merge/bi-arch confinement library paths
 - Fix confinement support for "nvidia-arch" GL functionality
 - Support Solus's GL "linux-driver-management" layout
 - Fix bind-mount confinement issues for snapd directory

Signed-off-by: Ikey Doherty <ikey@solus-project.com>

zyga commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -22,6 +28,18 @@
+    /usr/lib64/libcap.so* mr,
+    /usr/lib64/libcgmanager.so* mr,
+    /usr/lib64/libdl-[0-9]*.so* mr,
+    /usr/lib64/libnih.so* mr,
```
libnih is I think only used on 14.04 systems, you may not need those unless they are really present on solus.

ikeydoherty replied:
I'll drop nih

zyga commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -347,6 +376,25 @@
+        /usr/lib64/libgcc_s.so* mr,
+
+        # normal libs in order biarch + usr-merged
+        /usr/lib64/libapparmor.so* mr,
```
Same comment as above (may need pruning)

zyga commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -154,13 +172,21 @@
 
     mount options=(rw rbind) /usr/src/ -> /tmp/snap.rootfs_*/usr/src/,
     mount options=(rw rslave) -> /tmp/snap.rootfs_*/usr/src/,
+
```
I'm curious about those two rules, they look fine but I wasn't aware we need them or that we perform those operations.

zyga commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -22,6 +28,15 @@
+    /usr/lib64/libdbus-1.so* mr,
+    /usr/lib64/libudev.so* mr,
+    /usr/lib64/libseccomp.so* mr,
+    /usr/lib64/libseccomp.so* mr,
```
Duplicate

zyga commented on cmd/autogen.sh:
```diff
@@ -45,6 +45,9 @@ case "$ID" in
 		extra_opts="--libexecdir=/usr/lib/snapd --disable-apparmor"
 		;;
+	solus)
+		extra_opts="--with-snap-mount-dir=/snap --enable-merged-usr --enable-nvidia-arch $@"
```
Curious about the `$@` I didn't notice it on my first read.

ikeydoherty replied:
In the Solus package we do:

`%autogen` which actually calls `./autogen.sh %CONFOPTS%` which passes all the configure arguments along, this makes `autogen.sh` complete the chain and pass them along to `configure` so that the whole thing is configured properly in one go

Conan-Kudo replied:
Does anyone actually use this script in packaging? AFAIK, it was originally created for the test harness...

ikeydoherty replied:
Yeah - it lets you configure the snap confine part, I'm using that within the package.yml ...

ikeydoherty commented:
I'd agree that as the Solus support gets fleshed out that we might want that stuff to be conditional to save up spamming the rules/filters, and once we're at the multilib cross-roads we might want some Solus helper code to correctly append to the rules

niemeyer commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -11,6 +11,12 @@
+    /usr/lib64/libc{,-[0-9]*}.so* mr,
+    /usr/lib64/libpthread{,-[0-9]*}.so* mr,
+    /usr/lib64/librt{,-[0-9]*}.so* mr,
+    /usr/lib64/libgcc_s.so* mr,
```
Let's please get a pass by @jdstrand on these rules.

niemeyer commented on interfaces/builtin/opengl.go:
```diff
@@ -35,6 +35,12 @@ const openglConnectedPlugAppArmor = `
+  /var/lib/snapd/hostfs/usr/lib*/glx-provider/** rm,
+  /var/lib/snapd/hostfs/usr/lib*/libcuda* rm,
+  /var/lib/snapd/hostfs/usr/lib*/libn* rm,
+  /var/lib/snapd/hostfs/usr/lib*/lib{GL,EGL}* rm,
```
Why is that necessary now and wasn't necessary before? @zyga, @mvo?

Note we have some logic under cmd/snap-confine/mount-support-nvidia.c related to this.

ikeydoherty replied:
Well the comment in the context does explain that, its required for Solus NVIDIA libraries, which are accessed under /usr/lib{,64}/glx-provider due to the /usr/lib*/libGL type symlinks, and the rest of the libs (actual NVIDIA ones, not conflicting OpenGL ones) live in the main library directory. And the non-Solus part was never present because the rules were created for Arch Linux which doesn't actually have full apparmor confinement, so it was never exposed until now.

zyga replied:
Exactly as @ikeydoherty explained. In solus nvidia layout is similar to arch so snap-confine was built with `--with-nvidia-arch` which was confined to begin with.

niemeyer replied:
@zyga Sorry for being a bit slow, here but the question is this: how are people able to use nvidia libraries on Ubuntu if we don't have those rules, and where is the rule in the nvidia support of snap-confine that puts those libraries into /var/lib/snapd/lib/gl as they should be?

jdstrand replied:
I think we want something more like this:
```
# Solus LDM locations
/var/lib/snapd/hostfs/usr/lib{,32,64}/glx-provider/**.so{,.*} rm,
￼/var/lib/snapd/hostfs/usr/lib{,32,64}/libcuda*.so{,.*} rm,
￼/var/lib/snapd/hostfs/usr/lib{,32,64}/libnvidia*.so{,.*} rm,
￼/var/lib/snapd/hostfs/usr/lib{,32,64}/lib{GL,EGL}.so{,.*} rm,
```

Can you test and report back any errors? I thought using `lib{,32,64}` is good for future-proofing, but I also wanted to tighten up the rules to only mmap() .so files. Also, libn* seemed too generic so I guessed you wanted libnvidia*, but please adjust as necessary. If the above doesn't work for you, feel free to adjust and/or paste the apparmor denials.

ikeydoherty replied:
Well libnv maybe, there are a few in there but yeah we'll tighten them up and allow the biarch approach (We do have the lib32 dirs too fwiw)

zyga replied:
There are two implementations of nvidia support, one for "ubuntu" like systems and one for "arch" like systems. The ubuntu part had the correct apparmor rules because we had to write those. The arch part never had because there was no good place to actually explore and test those. Solus is "arch" like and with enabled apparmor was the first distribution to explore this combination.

zyga replied:
Note that on "arch" like systems /var/lib/snapd/lib/gl contains a tmpfs with a symlink farm so the effective files we access are different from what we did on Ubuntu.

niemeyer commented on mkversion.sh:
```diff
@@ -32,6 +32,12 @@ if [ ! -z "$1" ]; then
     o=shell
 fi
 
+# If the version is passed in the environment during build, let's use that
```
That doesn't look right. It implies the logic that is telling the code what its version is will be outside the code itself, which means there's a very realistic probability of updating the code without changing the version. That will create a lot of very confusing issues, including serious breakages as we do take the version into account in some places.

The right way of doing this is running mkversion.sh when the code is extracted from git. The result will then have the version encoded statically and may be shipped anywhere without the git metadata.

ikeydoherty replied:
Yes but this isn't being built from git, it's being built from the vendor tarball (which apparently doesn't know its own version) - thus passing the version in from the package metadata during build

zyga replied:
@niemeyer this matches what we do elsewhere but since we don't want to force people to run dpkg-parsechangelogs we offer a simple "this is the version right here" already (via command line argument). This patch extends the feature to environment value since it is easier to integrate in souls.

Conan-Kudo replied:
I don't know if it's a good idea for mkversion to just be pulling random shell variables. I added the ability to pass it as a parameter so that you can control the input.

ikeydoherty replied:
But `mkversion` is within the `autogen` pipeline, so to support correct and proper `autogen` invocation, the call to mkversion has to "learn" it from somewhere.. I also wouldn't really call it "random" given that its part of the documented ypkg build environment, which is sanely controlled. If it's gonna be that much of a problem I can split it to call mkversion **prior** to the autogen call, but I'd still need the parameters passed to configure like I'm doing now (Otherwise, it's a pretty pointless autogen)

Conan-Kudo replied:
I suggest calling it before the autogen call, as I do for the Fedora packaging.

niemeyer replied:
I don't think we want environment variable, positional parameters, or flag parameters that provide a version into this script. We want this script to tell the version based on the git data, and we wan this version to be statically saved into the filesystem so that it can be packed. If we have vendor tarballs that do not contain their own version today, these tarballs are broken and should be fixed. We want the source code to know its version, instead of expecting the person that builds the tarball to keep its version in mind and pass it in via a parameter. That latter approach will **definitely result in binaries reporting their version incorrectly eventually**.

niemeyer replied:
Let's please fix any such broken tarballs now, while it's early and we haven't suffered any major pain due to that.

ikeydoherty commented:
I've rebased the PR on master (dropping the commits/merges added on top of the commit) - and removed the `mkversion` changes, we'll invoke that in packaging prior to calling `%autogen`

jdstrand commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -235,6 +251,9 @@
     mount options=(rw bind) /usr/lib/nvidia-*/ -> /{tmp/snap.rootfs_*/,}var/lib/snapd/lib/gl/,
+    /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/* w,
+    mount fstype=tmpfs none -> /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/,
+    mount options=(remount,ro) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/,
```
For consistency with other mount rules in the policy, can you change this rule of yours to be this equivalent rule (options may be space or comma separated):
```
mount options=(remount ro) -> /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/,
```

jdstrand commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -347,6 +366,21 @@
+        /usr/lib64/libdbus-1.so* mr,
+        /usr/lib64/libudev.so* mr,
+        /usr/lib64/libseccomp.so* mr,
+
```
See above.

ikeydoherty commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -235,6 +251,9 @@
     /usr/** r,
     mount options=(rw bind) /usr/lib/nvidia-*/ -> /{tmp/snap.rootfs_*/,}var/lib/snapd/lib/gl/,
+    /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/* w,
+    mount fstype=tmpfs none -> /tmp/snap.rootfs_*/var/lib/snapd/lib/gl/,
```
That's what I originally had, which is why I stripped options from it (didn't work, I'd copy-edited the other tmpfs one i'd found)

jdstrand replied:
Can you paste the denial when it didn't work?

zyga commented on cmd/cmd.go:
```diff
@@ -67,7 +67,7 @@ func distroSupportsReExec() bool {
 	}
 	switch release.ReleaseInfo.ID {
-	case "fedora", "centos", "rhel", "opensuse", "suse", "poky", "arch":
+	case "fedora", "centos", "rhel", "opensuse", "suse", "poky", "arch", "solus":
```
I'll adjust this.

Conan-Kudo replied:
I don't know why we don't just have a list where re-exec is enabled rather than a disable list. The enable list is many times shorter (it's just Ubuntu and Debian).

ikeydoherty replied:
Thought the same thing tbh

jdstrand commented:
Note if you merge with upstream/master, you'll get https://github.com/snapcore/snapd/pull/3789 which will address my comments regarding future-proofing the policy. Other comments remain.

zyga commented:
I'll try to pick this up and get it merged.

niemeyer commented:
I've heard from @zyga today in the standup that this is being broken down into smaller pieces, so I'm closing this one.

@zyga, please keep us posted about the progress on this.
